### ISSUE TITLE
Stateful monitoring channels

### ISSUE THREAD

**Issue by goodfeli** [lisa-lab/pylearn2#173]
A lot of monitoring channels involve modifying some sort of state, either when you define the channel, or when you compute the channel, or both. Mostly this happens for channels involving a random number generator, such as the objective function for an MLP trained with dropout. Currently, the same MRG_RandomStreams object is used to generate the dropout masks for the TrainingAlgorithm and for computing the value of the objective on all monitoring datasets.

This is bad because it means the actual parameters a job runs depends on how many datasets are monitored. If you keep track of the training error as well as the validation set error, you will actually learn something different from if you just keep track of the validation set error.

We should figure out some way to cleanly use different state for each monitoring channel, so that the choice of what to monitor doesn't affect the outcome of the job.
